.clang-format

```diff
@@ -0,0 +1,96 @@
+---
+Language:        Cpp
+# BasedOnStyle:  LLVM
+AccessModifierOffset: -2
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: false
+AlignConsecutiveDeclarations: false
+AlignEscapedNewlinesLeft: false
+AlignOperands:   true
+AlignTrailingComments: true
+AllowAllParametersOfDeclarationOnNextLine: true
+AllowShortBlocksOnASingleLine: false
+AllowShortCaseLabelsOnASingleLine: false
+AllowShortFunctionsOnASingleLine: Inline
+AllowShortIfStatementsOnASingleLine: false
+AllowShortLoopsOnASingleLine: false
+AlwaysBreakAfterDefinitionReturnType: None
+AlwaysBreakAfterReturnType: None
+AlwaysBreakBeforeMultilineStrings: false
+AlwaysBreakTemplateDeclarations: true
+BinPackArguments: true
+BinPackParameters: false
+BraceWrapping:   
+  AfterClass:      true
+  AfterControlStatement: true
+  AfterEnum:       true
+  AfterFunction:   true
+  AfterNamespace:  true
+  AfterObjCDeclaration: true
+  AfterStruct:     true
+  AfterUnion:      true
+  BeforeCatch:     true
+  BeforeElse:      true
+  IndentBraces:    false
+BreakBeforeBinaryOperators: None
+BreakBeforeBraces: Allman
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializersBeforeComma: true
+BreakAfterJavaFieldAnnotations: false
+BreakStringLiterals: true
+ColumnLimit:     100
```

Comment:
I personally don't like setting `ColumnLimit`. In my opinion it restricts the freedom to linebreak for clarity too much since it always re-breaks everything.

Reply:
I added a break penalty to comments so it won't break those, for code I think somewhere between 80-120 should be used, otherwise it's a pain to read on a laptop screen. Even worse if looking at a side by side diff.

Reply:
Sure, I'm not advocating for 200-char lines. My problem is the following: If any `ColumnLimit` is set at all and I write e.g.

```
SomeFunctionThatTakesVarargs(
  OPTION_1, 1234,
  OPTION_2, 2345,
  OPTION_3, 3456,
  OPTION_END
);
```

it will rewrite it to

```
SomeFunctionThatTakesVarargs(OPTION_1, 1234, OPTION_2, 2345, OPTION_3, 3456, OPTION_END);
```

which I think is less clear (becomes more pronounced the more arguments you have).

Reply:
Then we should change the binpack arguments setting, it will either try to binpack them or spread them out nicely aligned on separate rows. I don't care that much either way as long as we have something that's consistent everywhere.
https://clang.llvm.org/docs/ClangFormatStyleOptions.html There's both `BinpackArguments` and `BinpackParameters`.

Reply:
I know. But then it will be

```
SomeFunctionThatTakesVarargs(
  OPTION_1,
  1234,
  OPTION_2,
  2345,
  OPTION_3,
  3456,
  OPTION_END
);
```

which I also feel is sub-par :-P

Reply:
hehe, how about not designing such a weird function taking pairs of arguments that aren't a pair/tuple/struct? 😃

Reply:
So we stop using 3rdparty libraries? ;-)

Reply:
hehe, I'd say yes if I encountered that method 😄 which library is that?
If it's a single use of a single library it's possible to use ` // clang-format off/on` for that file or section of the code.

Reply:
You get this with pretty much every C library that has functions with dynamic number of arguments. Where I encountered this specifically though is EGL. It takes the arguments as flat array instead of function parameters, but It's really the same problem.

Reply:
Example: https://dbus.freedesktop.org/doc/api/html/group__DBusMessage.html#ga591f3aab5dd2c87e56e05423c2a671d9

Reply:
ahh. That's not a pattern regularly used on Windows so not really familiar with it.
Not really sure how best to handle this as it would more or less disable some formatting that's nice to have in the rest of the code base.